src/main/scala/io/github/pauljamescleary/petstore/infrastructure/endpoint/UserEndpoints.scala

```diff
@@ -9,20 +11,49 @@ import org.http4s.dsl.Http4sDsl
 import org.http4s.{EntityDecoder, HttpService}
 
 import scala.language.higherKinds
-import io.github.pauljamescleary.petstore.domain.{UserAlreadyExistsError, UserNotFoundError}
-import io.github.pauljamescleary.petstore.domain.users.{User, UserService}
-
-class UserEndpoints[F[_]: Effect] extends Http4sDsl[F] {
+import domain._
+import domain.users._
+import domain.authentication._
+import tsec.passwordhashers.PasswordHash
+import tsec.passwordhashers.core.PasswordHasher
 
+class UserEndpoints[F[_]: Effect, A, K] extends Http4sDsl[F] {
   import Pagination._
   /* Jsonization of our User type */
-  implicit val userDecoder: EntityDecoder[F, User] = jsonOf[F, User]
 
-  private def signupEndpoint(userService: UserService[F]): HttpService[F] =
+  implicit val userDecoder: EntityDecoder[F, User] = jsonOf
+  implicit val loginReqDecoder: EntityDecoder[F, LoginRequest] = jsonOf
+
+  implicit val signupReqDecoder: EntityDecoder[F, SignupRequest] = jsonOf
+
+  private def loginEndpoint(userService: UserService[F], cryptService: PasswordHasher[A]): HttpService[F] =
+    HttpService[F] {
+      case req @ POST -> Root / "login" =>
+        val action: EitherT[F, UserAuthenticationFailedError, User] = for {
+          login <- EitherT.liftF(req.as[LoginRequest])
+          name = login.userName
+          user <- userService.getUserByName(name).leftMap(_ => UserAuthenticationFailedError(name))
+          valid <- EitherT.liftF(cryptService.checkpw(login.password, PasswordHash[A](user.hash)))
+          resp <-
+            if(valid) EitherT.rightT[F, UserAuthenticationFailedError](user)
+            else EitherT.leftT[F, User](UserAuthenticationFailedError(name))
+        } yield resp
+
+        action
```

Comment:
This could be the following to stay consistent with the other endpoints.  I don't love the `value` thing, but I think it is better to be consistent...


```
        action.value.flatMap {
          case Right(u) => Ok(u.asJson)
          case Left(UserAuthenticationFailedError(name)) =>
            BadRequest(s"Authentication failed for user $name")
        }
```